### PR TITLE
Add TDS token data length bounds checks

### PR DESCRIPTION
**Pull request by paulmedynski**
## Summary

Adds bounds checks on data lengths read from the TDS wire to prevent unbounded `byte[]` allocation from a malicious server.

All checks throw `SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, ...)` when a spoofed length exceeds protocol-reasonable maximums.

## Changes

| File | Change | Description | Tests |
|------|--------|-------------|-------|
| `TdsParser.cs` L4180 | `TryProcessSessionState` — total length check | Rejects `tokenLen > MaxTokenDataLength` (1 MB) | `SessionState_OversizedTotalLength` |
| `TdsParser.cs` L4233 | `TryProcessSessionState` — inner option length check | Rejects individual state option with `stateLen < 0` or `stateLen > MaxTokenDataLength` | `SessionState_OversizedInnerOptionLength`, `SessionState_NegativeInnerOptionLength` |
| `TdsParser.cs` L4454 | `TryProcessFedAuthInfo` — token length check | Rejects `tokenLen > MaxTokenDataLength` | `FedAuthInfo_OversizedTokenLength` |
| `TdsParser.cs` L3351 | `TryProcessEnvChange` — PromoteTransaction newLength check | Rejects `newLength > MaxPromoteTransactionLength` (64 KB) | `EnvChange_PromoteTransaction_OversizedLength` (login), `BatchResponse_PromoteTransaction_OversizedLength` (post-login) |
| `TdsParser.cs` L3853 | `TryProcessFeatureExtAck` — data length check | Rejects `dataLen > MaxTokenDataLength` | `FeatureExtAck_OversizedDataLength`, `FeatureExtAck_MaxAllowedDataLength_DoesNotThrow` |
| `TdsParser.cs` L7144 | `TryReadSqlDateTime` — data length check | Rejects `length > MaxDateTimeLength` (10) | `BatchResponse_DateTime_OversizedLength` |
| `TdsParser.cs` L4937-4948 | `TryProcessReturnValue` — non-PLP length check | Restructured: `IsPlp` types skip bounds check; non-PLP rejects `valLen > int.MaxValue` | `BatchResponse_ReturnValue_OversizedLength`, `BatchResponse_ReturnValue_PlpUnknownLength` |
| `TdsParser.cs` L7405 | `TryReadSqlValueInternal` — binary/vector length check | Rejects `length < 0` or `length > MAXSIZE` (8000) for binary types in sql_variant path | `BatchResponse_SqlVariantBinary_OversizedLength`, `BatchResponse_SqlVariantBinary_NegativeLength` |
| `SqlConnectionInternal.cs` L1349 | `OnFeatureExtAck` — SRECOVERY inner state length | Validates inner state option length doesn't exceed remaining buffer | `SRecovery_MalformedInnerStateLength` |
| `TdsEnums.cs` L90-96 | Constants | `MaxTokenDataLength` (1 MB), `MaxPromoteTransactionLength` (64 KB), `MaxDateTimeLength` (10) | *(used by all above)* |

### Group 1: New upper-bound constraints

These add hard limits where previously the parser would attempt to allocate whatever the server claimed. A **conforming SQL Server** will never exceed any of these limits, but a misbehaving proxy, network appliance, or third-party TDS implementation could.

| Location | Limit | Rationale |
|----------|-------|-----------|
| FeatureExtAck `dataLen` | 1 MB | No legitimate feature ack payload approaches this |
| SessionState total length | 1 MB | Session state tokens are typically single-digit KB |
| SessionState inner option | ≥ 0 and ≤ 1 MB | Individual state options are even smaller; negative values indicate wire corruption |
| FedAuthInfo `tokenLen` | 1 MB | Contains only STS URL + SPN — a few hundred bytes |
| PromoteTransaction `newLength` | 64 KB | DTC propagation tokens are typically ~200–500 bytes |
| DateTime `length` | 10 bytes | DateTimeOffset (the largest) is exactly 10; this is the protocol ceiling |
| Binary/vector in sql_variant | ≥ 0 and ≤ 8000 | MAXSIZE is the protocol maximum for non-PLP binary; negative `lenData` results from a total variant length smaller than the type overhead |

**Risk to existing apps:** Effectively zero against real SQL Server. Could break connections through a non-conforming TDS proxy that inflates token lengths (e.g., for padding or encapsulation).

### Group 2: Debug.Assert → runtime exception (bug fix)

These replaced `Debug.Assert` (which is stripped in Release builds, meaning the invalid state was silently accepted) with a throwing check:

| Location | Previous behavior (Release) | New behavior |
|----------|----------------------------|--------------| 
| ReturnValue non-PLP `valLen > int.MaxValue` | Silently truncated to `int.MaxValue` | Throws `ParsingErrorLength` |
| Binary/Vector in sql_variant `length > MAXSIZE` | No check at all (assert stripped) | Throws `ParsingErrorLength` |

**Risk to existing apps:** Only if the app was previously "surviving" corrupt data by silently reading garbage. In practice these indicate a corrupted stream that would likely cause downstream failures anyway.

### Group 3: Buffer overrun prevention (defense-in-depth)

| Location | Issue prevented |
|----------|-----------------|
| `SqlConnectionInternal.OnFeatureExtAck` SRECOVERY: `len < 0 \|\| len > data.Length - i` | Out-of-bounds `Buffer.BlockCopy` from a malformed feature ack payload |

**Risk to existing apps:** None — this path only fires if the server sends internally inconsistent SRECOVERY data (claimed inner length exceeds the outer buffer already validated by the FeatureExtAck check).

### Group 4: Behavioral improvement (non-breaking)

| Location | Change |
|----------|--------|
| `TryReadSqlDateTime` | Replaced `((uint)length <= 16) ? stackalloc : new byte[length]` with unconditional `stackalloc byte[10]` |

Since the bounds check guarantees `length ≤ 10`, the conditional heap allocation path is dead code eliminated. Minor perf win. **No user-visible behavior change.**

### Impact summary

| Category | Count | Risk to legitimate apps |
|----------|-------|------------------------|
| New constraints | 6 sites | ~Zero (limits far exceed protocol maximums) |
| Assert → exception | 2 sites | Near-zero (previously silent corruption) |
| Buffer overrun fix | 1 site | None |
| Perf improvement | 1 site | None |

The only scenario where a real application could be affected is if it connects through a non-conforming TDS intermediary (proxy/load balancer) that rewrites token lengths to values exceeding these bounds. Standard SQL Server, Azure SQL, and conforming TDS 7.x/8.0 implementations will never produce values exceeding any of these limits.

## Test infrastructure

- Added `OnSQLBatchCompleted` hook to `GenericTdsServer` for post-login TDS token injection
- Added `OnAuthenticationResponseCompleted` hook usage for login-phase token injection
- Per-test instance fixtures (`IDisposable`) for proper test isolation — each test gets a fresh `TdsServerFixture`
- 14 unit tests covering all bounds checks (login-phase, post-login, positive/negative boundary conditions)

## Checklist

- [x] Tests added or updated
- [x] No breaking changes introduced
- [x] PLP regression verified (test 9 confirms valid PLP sentinel is not rejected)
- [x] 100% patch coverage on new bounds checks

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs [hunk 1]
```diff
@@ -1346,6 +1346,11 @@ internal void OnFeatureExtAck(int featureId, byte[] data)
                             len = bLen;
                         }
 
+                        if (len < 0 || len > data.Length - i)
+                        {
+                            throw SQL.ParsingError(ParsingErrorState.CorruptedTdsStream);
+                        }
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs:1352]: Fixed in 54a28178a. Changed to `SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, len)` to include the offending length, consistent with other bounds checks in this PR.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/TdsTokenBoundsTests.cs [hunk 2]
```diff
@@ -0,0 +1,844 @@
...
+    // ──────────────────────────────────────────────────────────────────────────
+    // Test 7: Post-login batch response — DateTime oversized length
+    // ──────────────────────────────────────────────────────────────────────────
+
+    /// <summary>
+    /// Verifies that <c>TryReadSqlDateTime</c> rejects a TIME column value whose
+    /// data length byte exceeds the maximum datetime wire size (10 bytes). A
+    /// malicious server could set this length to a large value, causing the
+    /// parser to attempt an oversized stackalloc.
+    /// </summary>
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/TdsTokenBoundsTests.cs:514]: Fixed in 54a28178a. Updated comment from "oversized stackalloc" to "unbounded heap allocation".

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsEnums.cs [hunk 3]
```diff
@@ -84,6 +84,14 @@ internal static class TdsEnums
         public const int MAX_PACKET_SIZE = 32768;
         public const int MAX_SERVER_USER_NAME = 256;  // obtained from luxor
 
+        // Maximum allowed data length for login-phase token payloads (feature ext ack,
+        // session state, fedauth info). Prevents a malicious server from causing
+        // unbounded memory allocation on the client before authentication completes.
+        internal const int MaxTokenDataLength = 1 << 20; // 1 MB
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsEnums.cs]: Fixed in 54a28178a. Broadened comment to: "Maximum data length for feature/session tokens. Bounds-checks using this constant reject payloads that would cause unbounded heap allocations from a spoofed length field."

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 4]
```diff
@@ -4912,14 +4932,20 @@ internal TdsOperationStatus TryProcessReturnValue(int length,
...
+            int intlen;
 
             if (rec.metaType.IsPlp)
             {
                 intlen = int.MaxValue;    // If plp data, read it all
             }
+            else if (valLen > (ulong)int.MaxValue)
+            {
+                throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, int.MaxValue);
+            }
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:4945]: Fixed in 54a28178a. Now uses `unchecked((int)valLen)` which reports the actual on-wire value (e.g. `-1` for `0xFFFFFFFF`) instead of the misleading `int.MaxValue`.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/FeatureExtAckBoundsTests.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    /// off-by-one in the bounds check.
+    /// </summary>
+    [Fact]
+    public void FeatureExtAck_MaxAllowedDataLength_DoesNotThrow()
+    {
+        // Arrange: inject a FeatureExtAck token at exactly the maximum allowed size.
+        // We use a small real payload to avoid actually allocating 1 MB in the test;
+        // the server will write a data length that equals the actual data length.
+        // This verifies that the boundary value is accepted (not off-by-one).
+        byte[] legitimateData = new byte[] { 0x01 }; // 1-byte GlobalTransactions ack
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/FeatureExtAckBoundsTests.cs]: Fixed in 54a28178a. Rewrote the test to inject a `MaliciousFeatureExtAckToken` with `claimedDataLen = MaxTokenDataLength` (exactly at the boundary). The parser accepts the length (bounds check doesn't fire), then fails for other reasons trying to read that much data. The test asserts the error is NOT the bounds-check state-18 error.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsEnums.cs [hunk 6]
```diff
@@ -84,6 +84,14 @@ internal static class TdsEnums
...
+
+        // Maximum allowed data length for a DTC promote transaction propagation token.
+        internal const int MaxPromoteTransactionLength = 1 << 16; // 64 KB
+
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsEnums.cs:94]: Fixed in 54a28178a. Added `internal const int MaxDateTimeLength = 10;` in `TdsEnums.cs` and updated `TryReadSqlDateTime` to reference it instead of using a local const.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 7]
```diff
@@ -3846,10 +3850,14 @@ private TdsOperationStatus TryProcessFeatureExtAck(TdsParserStateObject stateObj
                     {
                         return result;
                     }
+                    if (dataLen > (uint)TdsEnums.MaxTokenDataLength)
+                    {
+                        throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, (int)dataLen);
+                    }
                     byte[] data = new byte[dataLen];
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs]: Fixed in 9f4bcf0e8. Introduced a local `int dataLength = (int)dataLen` after the bounds check, used for both the array allocation and `TryReadByteArray` call — eliminates the repeated cast.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/FeatureExtAckBoundsTests.cs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+        // Act & Assert: connection should fail with a parsing error, NOT an OutOfMemoryException
+        using SqlConnection connection = new(_connectionString);
+        Exception ex = Assert.ThrowsAny<InvalidOperationException>(connection.Open);
+
+        // The exception message should indicate a corrupted TDS stream (parsing error state 18)
+        // with the oversized length value, not an OOM from attempting the allocation.
+        Assert.Contains("18", ex.Message); // ParsingErrorState.CorruptedTdsStream = 18
+        Assert.Contains((TdsEnums.MaxTokenDataLength + 1).ToString(), ex.Message);
+    }
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/FeatureExtAckBoundsTests.cs:84]: Fixed in 9f4bcf0e8. Wrapped the test body in `try/finally` with `_server.OnAuthenticationResponseCompleted = null` in the `finally` block, matching the pattern used by the other tests.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 9]
```diff
@@ -4218,6 +4231,10 @@ private TdsOperationStatus TryProcessSessionState(TdsParserStateObject stateObj,
                         return result;
                     }
                 }
+                if (stateLen < 0 || stateLen > TdsEnums.MaxTokenDataLength)
+                {
+                    throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, stateLen);
+                }
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:4237]: **Test coverage:** `TdsTokenBoundsTests.SessionState_OversizedInnerOptionLength_ThrowsParsingError` — injects a SessionState token with a valid outer length but an inner state option whose 0xFF + DWORD encoded length exceeds MaxTokenDataLength.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 10]
```diff
@@ -4169,6 +4178,10 @@ private TdsOperationStatus TryProcessSessionState(TdsParserStateObject stateObj,
             {
                 throw SQL.ParsingErrorLength(ParsingErrorState.SessionStateLengthTooShort, length);
             }
+            if (length > TdsEnums.MaxTokenDataLength)
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:4181]: **Test coverage:** `TdsTokenBoundsTests.SessionState_OversizedTotalLength_ThrowsParsingError` — injects a SessionState token (0xE4) whose total length field exceeds MaxTokenDataLength.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 11]
```diff
@@ -3348,6 +3348,10 @@ private TdsOperationStatus TryProcessEnvChange(int tokenLength, TdsParserStateOb
                             // new value has 4 byte length
                             return result;
                         }
+                        if (env._newLength < 0 || env._newLength > TdsEnums.MaxPromoteTransactionLength)
+                        {
+                            throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, env._newLength);
+                        }
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:3354]: **Test coverage:** `TdsTokenBoundsTests.EnvChange_PromoteTransaction_OversizedLength_ThrowsParsingError` and `TdsTokenBoundsTests.BatchResponse_PromoteTransaction_OversizedLength_ThrowsParsingError` — the first injects during login, the second during command execution via `OnSQLBatchCompleted`.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 12]
```diff
@@ -3846,10 +3850,15 @@ private TdsOperationStatus TryProcessFeatureExtAck(TdsParserStateObject stateObj
...
-                    byte[] data = new byte[dataLen];
-                    if (dataLen > 0)
+                    if (dataLen > (uint)TdsEnums.MaxTokenDataLength)
+                    {
+                        throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, (int)dataLen);
+                    }
+                    int dataLength = (int)dataLen;
+                    byte[] data = new byte[dataLength];
+                    if (dataLength > 0)
                     {
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:3860]: **Test coverage:** `FeatureExtAckBoundsTests.FeatureExtAck_OversizedDataLength_ThrowsParsingError` (injects dataLen = MaxTokenDataLength + 1, asserts parsing error state 18) and `FeatureExtAckBoundsTests.FeatureExtAck_MaxAllowedDataLength_DoesNotThrow` (boundary test confirming no off-by-one).

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 13]
```diff
@@ -4435,6 +4452,10 @@ private TdsOperationStatus TryProcessFedAuthInfo(TdsParserStateObject stateObj,
                 SqlClientEventSource.Log.TryTraceEvent("<sc.TdsParser.TryProcessFedAuthInfo|ERR> FEDAUTHINFO token stream length too short for CountOfInfoIDs.");
                 throw SQL.ParsingErrorLength(ParsingErrorState.FedAuthInfoLengthTooShortForCountOfInfoIds, tokenLen);
             }
+            if (tokenLen > TdsEnums.MaxTokenDataLength)
+            {
+                throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, tokenLen);
+            }
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:4458]: **Test coverage:** `TdsTokenBoundsTests.FedAuthInfo_OversizedTokenLength_ThrowsParsingError` — injects a FedAuthInfo token (0xEE) with tokenLen exceeding MaxTokenDataLength. The token type is dispatched unconditionally by TryRun.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 14]
```diff
@@ -4912,14 +4933,20 @@ internal TdsOperationStatus TryProcessReturnValue(int length,
...
             {
                 intlen = int.MaxValue;    // If plp data, read it all
             }
+            else if (valLen > (ulong)int.MaxValue)
+            {
+                throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, unchecked((int)valLen));
+            }
+            else
+            {
+                intlen = (int)valLen;
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:4948]: **Test coverage:** `TdsTokenBoundsTests.BatchResponse_ReturnValue_OversizedLength_ThrowsParsingError` (injects IMAGE return value with data length -1, which as ulong exceeds int.MaxValue) and `TdsTokenBoundsTests.BatchResponse_ReturnValue_PlpUnknownLength_Succeeds` (regression guard: PLP unknown-length sentinel 0xFFFFFFFFFFFFFFFE must NOT trigger the check).

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 15]
```diff
@@ -7113,7 +7140,14 @@ internal TdsOperationStatus TryReadSqlValue(SqlBuffer value,
...
+            // Reject anything larger to prevent heap allocation from spoofed metadata.
+            if (length < 0 || length > TdsEnums.MaxDateTimeLength)
+            {
+                throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, length);
+            }
+
+            Span<byte> datetimeBuffer = stackalloc byte[TdsEnums.MaxDateTimeLength];
 
             TdsOperationStatus result = stateObj.TryReadByteArray(datetimeBuffer, length);
             if (result != TdsOperationStatus.Done)
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:7157]: **Test coverage:** `TdsTokenBoundsTests.BatchResponse_DateTime_OversizedLength_ThrowsParsingError` — injects a TIME column whose data length byte is 11 (exceeding MaxDateTimeLength = 10). Also validates the elimination of the conditional heap allocation: since all valid datetime types fit in ≤ 10 bytes, this is now an unconditional `stackalloc`.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs [hunk 16]
```diff
@@ -1346,6 +1346,11 @@ internal void OnFeatureExtAck(int featureId, byte[] data)
                             len = bLen;
                         }
 
+                        if (len < 0 || len > data.Length - i)
+                        {
+                            throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, len);
+                        }
+
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Connection/SqlConnectionInternal.cs:1353]: **Test coverage:** `TdsTokenBoundsTests.SRecovery_MalformedInnerStateLength_ThrowsParsingError` — injects a FeatureExtAck with SessionRecovery feature (ID=1) whose inner state data contains an option claiming length 999 while only 1 byte follows. The bounds check `len < 0 || len > data.Length - i` catches this overflow.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 17]
```diff
@@ -7369,7 +7403,10 @@ internal TdsOperationStatus TryReadSqlValueInternal(SqlBuffer value, byte tdsTyp
                 case TdsEnums.SQLVECTOR:
                     {
                         // Note: Better not come here with plp data!!
-                        Debug.Assert(length <= TdsEnums.MAXSIZE);
+                        if (length < 0 || length > TdsEnums.MAXSIZE)
+                        {
+                            throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, length);
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:7412]: **Test coverage:** `TdsTokenBoundsTests.BatchResponse_SqlVariantBinary_OversizedLength_ThrowsParsingError` — injects a sql_variant ROW containing a BigVarBinary inner type whose computed `lenData` (8001) exceeds MAXSIZE (8000). This is the only path to reach this check since normal binary columns go through `TryReadSqlValue` directly.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsEnums.cs [hunk 18]
```diff
@@ -84,6 +84,17 @@ internal static class TdsEnums
...
+        // unbounded memory allocation via spoofed token length fields.
+        internal const int MaxTokenDataLength = 1 << 20; // 1 MB
+
+        // Maximum allowed data length for a DTC promote transaction propagation token.
+        internal const int MaxPromoteTransactionLength = 1 << 16; // 64 KB
+
+        // Maximum valid wire size for datetime types (DateTimeOffset = 5 time + 3 date + 2 offset).
+        internal const int MaxDateTimeLength = 10;
+
         // Severity  0 - 10 indicates informational (non-error) messages
```
Comment (paulmedynski) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsEnums.cs:98]: **Test coverage:** These constants are validated by all 12 unit tests across `FeatureExtAckBoundsTests` (2 tests) and `TdsTokenBoundsTests` (10 tests). Each test verifies that crossing the respective threshold triggers `ParsingErrorState.CorruptedTdsStream`. The boundary test (`MaxAllowedDataLength_DoesNotThrow`) also confirms no off-by-one in the MaxTokenDataLength constant.

**Review by paulmedynski — COMMENTED** [review]
**Self-review: Test coverage annotations**

Each bounds check added to driver code has been annotated with the specific test(s) that exercise it. Summary:

| Driver change | Test(s) |
|---|---|
| `TdsEnums.cs` — MaxTokenDataLength, MaxPromoteTransactionLength, MaxDateTimeLength | All 12 tests validate these thresholds |
| `TdsParser.cs` — PromoteTransaction bounds check (L3351) | `EnvChange_PromoteTransaction_OversizedLength_ThrowsParsingError`, `BatchResponse_PromoteTransaction_OversizedLength_ThrowsParsingError` |
| `TdsParser.cs` — FeatureExtAck bounds check (L3853) | `FeatureExtAck_OversizedDataLength_ThrowsParsingError`, `FeatureExtAck_MaxAllowedDataLength_DoesNotThrow` |
| `TdsParser.cs` — SessionState total length (L4178) | `SessionState_OversizedTotalLength_ThrowsParsingError` |
| `TdsParser.cs` — SessionState inner option (L4234) | `SessionState_OversizedInnerOptionLength_ThrowsParsingError` |
| `TdsParser.cs` — FedAuthInfo token length (L4455) | `FedAuthInfo_OversizedTokenLength_ThrowsParsingError` |
| `TdsParser.cs` — ReturnValue non-PLP length (L4936) | `BatchResponse_ReturnValue_OversizedLength_ThrowsParsingError`, `BatchResponse_ReturnValue_PlpUnknownLength_Succeeds` |
| `TdsParser.cs` — DateTime length (L7143) | `BatchResponse_DateTime_OversizedLength_ThrowsParsingError` |
| `TdsParser.cs` — Binary/Vector sql_variant (L7405) | `BatchResponse_SqlVariantBinary_OversizedLength_ThrowsParsingError` |
| `SqlConnectionInternal.cs` — SRECOVERY inner state (L1349) | `SRecovery_MalformedInnerStateLength_ThrowsParsingError` |

All 12 tests pass in ~19s.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 19]
```diff
@@ -3846,10 +3850,15 @@ private TdsOperationStatus TryProcessFeatureExtAck(TdsParserStateObject stateObj
                     {
                         return result;
                     }
-                    byte[] data = new byte[dataLen];
-                    if (dataLen > 0)
+                    if (dataLen > (uint)TdsEnums.MaxTokenDataLength)
+                    {
+                        throw SQL.ParsingErrorLength(ParsingErrorState.CorruptedTdsStream, (int)dataLen);
+                    }
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParser.cs:3856]: Fixed in 3bb59e9. Added a `uint` overload of `SQL.ParsingErrorLength()` so the raw `uint dataLen` is passed directly without truncation. The format string renders it as an unsigned value, so wire lengths > `int.MaxValue` now display correctly (e.g. `4294967295` instead of `-1`).

File: src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/TdsTokenBoundsTests.cs [hunk 20]
```diff
@@ -0,0 +1,1031 @@
...
+            _listeners = new System.Diagnostics.TraceListener[System.Diagnostics.Trace.Listeners.Count];
+            System.Diagnostics.Trace.Listeners.CopyTo(_listeners, 0);
+            System.Diagnostics.Trace.Listeners.Clear();
+        }
+
+        public void Dispose()
+        {
+            System.Diagnostics.Trace.Listeners.AddRange(_listeners);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/TdsTokenBoundsTests.cs:1050]: Fixed in 3bb59e9. `DebugAssertSuppressor` now holds a static lock (`Monitor.Enter`/`Exit`) for its entire lifetime, serializing any concurrent access to the global `Trace.Listeners` collection. `Dispose` also clears listeners before restoring the saved set, preventing duplicates if something else modified the collection in the interim.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/TdsTokenBoundsTests.cs [hunk 21]
```diff
@@ -0,0 +1,1037 @@
...
+            System.Diagnostics.Trace.Listeners.CopyTo(_listeners, 0);
+            System.Diagnostics.Trace.Listeners.Clear();
+        }
+
+        public void Dispose()
+        {
+            System.Diagnostics.Trace.Listeners.Clear();
+            System.Diagnostics.Trace.Listeners.AddRange(_listeners);
+            System.Threading.Monitor.Exit(s_listenerLock);
+        }
```
[the remark this answers is not in the document]
Comment (paulmedynski) [reply src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/TdsTokenBoundsTests.cs:1049]: Fixed in 1918e49. Constructor now wraps the listener snapshot in `try/catch` — if allocation or `CopyTo` throws, `Monitor.Exit` is called before re-throwing. `Dispose` wraps the restore in `try/finally` so the lock is always released even if `AddRange` fails.

File: src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/ConnectionEnhancedRoutingTests.cs [hunk 22]
```diff
@@ -14,6 +14,8 @@ namespace Microsoft.Data.SqlClient.UnitTests.SimulatedServerTests;
 /// <summary>
 /// Tests connection routing using the enhanced routing feature extension and envchange token
 /// </summary>
+// TODO: Do we need this collection?  It serializes all tests within it, which we probably don't
```
Comment (mdaigle) [inline src/Microsoft.Data.SqlClient/tests/UnitTests/SimulatedServerTests/ConnectionEnhancedRoutingTests.cs:17]: At one point I tried to remove this, but there's some design flaw in the port assignment or cleanup that causes them to be recycled when there are still bytes on the wire. This resulted in test failures when later test runs picked up a dirty port. Maybe we're not doing enough to drain the data on the wire. Or maybe we're not properly shutting down our resources. I didn't have time to dig deeper and still haven't gotten back to it.

**Review by mdaigle — APPROVED** [review]

**Review by apoorvdeshmukh — APPROVED** [review]

**Review by cheenamalhotra — APPROVED** [review]
